lib/configgin.rb

```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
 
+    sts = kube_client_stateful_set.get_stateful_set(instance_group, kube_namespace)
+
+    # Make sure the secret attached to the stateful set exists.
+    # XXX This should probably be done by fissile via the helm chart.
     secret = Kubeclient::Resource.new
-    secret.metadata = {}
-    # Prefixing with pod.metadata.name is purely for human convenience/debugging.
-    secret.metadata.name = "#{pod.metadata.name}-#{pod.metadata.uid}"
-    secret.metadata.namespace = kube_namespace
-
-    # Make sure the secret gets removed when the pod is deleted.
-    secret.metadata.ownerReferences = [
-      {
-        apiVersion: pod.apiVersion,
-        blockOwnerDeletion: false,
-        controller: false,
-        kind: pod.kind,
-        name: pod.metadata.name,
-        uid: pod.metadata.uid,
-      }
-    ]
+    secret.metadata = {
+      name: sts.metadata.name,
+      namespace: kube_namespace,
+      ownerReferences:  [
+        {
+          apiVersion: sts.apiVersion,
+          blockOwnerDeletion: false,
+          controller: false,
+          kind: sts.kind,
+          name: sts.metadata.name,
+          uid: sts.metadata.uid,
+        }
+      ]
+    }
+    secret.data ||= {}
+    begin
+      kube_client.create_secret(secret)
+    rescue
+    end
+    secret = kube_client.get_secret(instance_group, kube_namespace)
+    secret.data ||= {}
+
+    version_tag = ENV["CONFIGGIN_VERSION_TAG"]
+    new_tag = !secret.data[version_tag]
+    if new_tag
+      secret.data = {version_tag => ""}
+    end
 
-    secret.data = {}
     digests = {}
     jobs.each do |name, job|
       digests[name] = property_digest(job.exported_properties)
       secret.data["skiff-exported-properties-#{name}"] = Base64.encode64(job.exported_properties.to_json)
-      secret.data["skiff-exported-digest-#{name}"] = Base64.encode64(digests[name])
-    end
 
-    # Only the main container gets to export properties; colocated sidecars don't.
-    if instance_group == ENV["KUBERNETES_CONTAINER_NAME"]
-      begin
-        kube_client.delete_secret(secret.metadata.name, kube_namespace)
-      rescue
+      encoded_digest = Base64.encode64(digests[name])
+      # XXX do we actually need these?
+      # secret.data["skiff-exported-digest-#{name}"] = encoded_digest
+
+      # Record initial digest values whenever the tag changes, in which case the pod startup
+      # order is already controlled by the "CONFIGGIN_IMPORT_#{role}" references to the new
+      # tags in the corresponding secrets. A missing import annotation reflects this set of
+      # exported properties because the sts as deployed by helm doesn't have any annotation
+      # and we don't want to trigger a redundant pod restart by adding them.
+      if new_tag
+        secret.data["skiff-initial-digest-#{name}"] = encoded_digest
+      end
+      if secret.data["skiff-initial-digest-#{name}"] == encoded_digest
+        digests[name] = nil
```

Comment:
So if the secret already has a matching digest (or if the tag is new), we'll delete the digest from the annotation?

Reply:
Yes, that is what the comment above was trying to explain: There is no annotation for the property values exported by an install or update, so pods don't need to restart due to the template change.